lms/services/application_instance.py

```diff
@@ -35,6 +36,85 @@ def __init__(self, application_instance: ApplicationInstance):
         )
 
 
+@dataclass
+class SearchBuilder:
+    base_model: BASE
+    clause_builders: dict
+    key_map: dict = field(default_factory=lambda: {"id_": "id", "type_": "type"})
+    query_start: Optional[callable] = None
+
+    def get_query(self, db_session, kwargs, combination=sa.and_):
+        query = self.query_start(db_session) or db_session.query(self.base_model)
```

Comment:
A sensible default to start a query which will be correct most of the time, with an escape hatch where required.